### DISCUSSION TITLE
Unknown Schemas on FastAPI Swagger

### DISCUSSION THREAD

**Discussion by igormorgado** [collerek/ormar#405] [Q&A] [answered]
There are some unkown schemas showing off in my FastAPI Swagger

![Screenshot from 2021-11-02 14-39-19](https://user-images.githubusercontent.com/2663394/139917487-5d96b902-518f-46d7-ba83-1a40a5928ee4.png)

Is there any way to hide them?

**Comment by igormorgado** [comment]
As far as I can tell, seems these fields are growing in numbers. What are those? Why they are being created? I


![Screenshot from 2021-11-03 09-24-05](https://user-images.githubusercontent.com/2663394/140059453-8035fca6-e8b2-4ff2-8189-b4e7965fb18c.png)

**Reply by collerek** [reply]
PKOnly models are auto created for you in order to allow you to issue a payload with a nested model with only primary key loaded, or allow you to return such models from a query. So if you have a model A that links to B (by fk), when you load only models A, ormar initialize for you a nested model of B with primary key only -> so you can later issue a query like `A.B.load()` to fetch the nested model.

That way you can pass and receive models like:
```
# pseudo json notation
A: { B: 1, id: 12} # B referenced by pk value
A: { B: {id: 1}, id: 12} # B is a primarykey only model - this is required to pass, as in "normal" circumstances other fields can be required (i.e. name) and would fail validation
A: { B: {id: 1 name: test}, id: 12} # B is fully populated model
```

The second screenshot - those are the models created by the `get_pydantic` as fastapi (and best practice) does not allow you to have multiple models with the same name coming from the same file/module (and `ormar` does not scan all files in your project so have no way of telling if given models are coming from the same file, so it applies to all pydantic models created from ormar models).

They are actually used in the response schema, so if I remove them it will probably blow up :)

**Reply by igormorgado** [reply]
Is possible to rename them on `get_pydantic()` call? That _XYZ naming isn't much helpful.

**Reply by igormorgado** [reply]
Reading the doc I could find that: 

Note that if you have nested models, it will generate whole tree of pydantic models for you!

That is probably there are so many of those are being created. Since my models are related with each other in many levels. For example

Asset -> Account (field name owner)
Asset -> Account (field name user)
Asset -> Contract -> Account (field name owner)
Asset -> Contract -> Account (field name user)
Account -> User

Hence we can some deep relationships as

Asset -> Contract -> Account(owner) -> User

**Reply by collerek** [reply]
Yes, as each time you can exclude different fields each call generates the new models. The XYZ is just a random 3 letters and it's not very readable, but it's hard to change it to anything meaningful. When you call `get_pydantic` the function does not know where do you call it from. If you generate multiple pydantic trees from same ormar models appending a parent/main model would not work as you would still have AccountAsset1, AccoountAsset2 etc.

**Reply by igormorgado** [reply]
If I take the approach to create manually a Pydantic model does that allow me to have a clean Schema?

**Reply by collerek** [reply]
If you use just `ormar` models it will be clear.
If you want to exclude fields and will reuse the same manually created models in different endpoints than it will be clear again.
If you exclude different fields in different endpoints than you still need to create as many models but you can rename them as you want.

**Reply by igormorgado** [reply]
I cannot tag you response as the answer.

**Reply by collerek** [reply]
copied the comment and marked as answered.

BTW you can probably manually modify the names of the models after creation by reassigning __name__. Didn;t check but should work.
But you cannot set the same names for two classes in one module as fastapi will crash.

**Comment by collerek** [comment, marked as answer]
If you use just ormar models it will be clear.
If you want to exclude fields and will reuse the same manually created models in different endpoints than it will be clear again.
If you exclude different fields in different endpoints than you still need to create as many models but you can rename them as you want.
